geode/src/main/scala/akka/stream/alpakka/geode/internal/stage/GeodeCQueryGraphLogic.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.geode.internal.stage
+
+import akka.stream.SourceShape
+import akka.stream.stage.StageLogging
+import org.apache.geode.cache.client.ClientCache
+import org.apache.geode.cache.query.{CqAttributesFactory, CqEvent, CqQuery, Struct}
+import org.apache.geode.cache.util.CqListenerAdapter
+
+abstract class GeodeCQueryGraphLogic[V](val shape: SourceShape[V],
+                                        val clientCache: ClientCache,
+                                        val queryName: Symbol,
+                                        val sql: String)
+    extends GeodeSourceStageLogic[V](shape, clientCache)
+    with StageLogging {
+
+  var query: Option[CqQuery] = None
+
+  override def preStart(): Unit = {
+
+    val cqf = new CqAttributesFactory()
+
+    val eventListener = new CqListenerAdapter() {
+      override def onEvent(ev: CqEvent): Unit =
+        onElement.invoke(ev.getNewValue().asInstanceOf[V])
+
+      override def onError(ev: CqEvent): Unit =
+        log.error(ev.getThrowable, s"$ev")
+
+      override def close(): Unit = {
+        log.debug("closes")
+        inFinish.invoke(())
+      }
+    }
+    cqf.addCqListener(eventListener)
+
+    val cqa = cqf.create()
+
+    query = Some(qs.newCq(queryName.name, sql, cqa))
+
+    query.foreach { q =>
+      handleQuery(q)
+    }
+    onConnect.invoke(())
+  }
+
+  private def handleQuery(q: CqQuery) = {
+    val res = q.executeWithInitialResults[Struct]()
+
+    val it = res.iterator()
+    while (it.hasNext) {
+      val str = it.next()
+      val v = str.getFieldValues()(1).asInstanceOf[V]
+      onElement.invoke(v)
+    }
```

Comment:
I see now why the test case for the continuous query was that style. There is actually no way to backpressure this `GeodeCQueryGraphLogic`. It starts and continues to emit elements to the stream from its preStart, which is not the best behaviour for a GraphStage.

Also why is `handleQuery` needed if above there is an `cqf.addCqListener(eventListener)` which also gets elements of type `V`?